### PR TITLE
feat: [NPM] ipset save before restoring and fixing grep UTs

### PR DESCRIPTION
**Pull request by huntergregory**
# Changes for Applying IPSets
- use ipset save file to add/delete members
- update error handling logic to skip previously run lines in a restore file

## Design Decision
Instead of calculating which members should be removed from ipsets, ipset manager will use the current kernel state via ipset save and apply the cache as the expected state. 

### Pros
- for the dirty sets, the kernel will always become exactly as we desire (delete everything not desired from the kernel i.e. no leftover set members that aren't in the cache)
- we are effectively reconciling and mitigating security concerns for dirty sets
- no need to modify cache management at this time

### Cons
- increased compute and latency 
- added complexity and perhaps harder to maintain

### Looking Forward
We will eventually have an infrequent reconcile routine which will reconcile the whole cache with the kernel state.

We will assess the extra latency of this design choice later.

# Other changes
- Allow proper exit codes and stdouts for piped commands in UTs.

TODO: 
- move some UT code into smaller-scoped functions for less complexity

### CODE REVIEW

File: npm/pkg/dataplane/ipsets/ipsetmanager_linux.go [hunk 1]
```diff
@@ -29,153 +67,409 @@ func (iMgr *IPSetManager) resetIPSets() error {
...
-			methodFlag = util.IpsetSetListFlag
-		} else if set.Type == NamedPorts {
-			methodFlag = util.IpsetIPPortHashFlag
-		}
+// updates the creator (adds/deletes members) for dirty sets already in the kernel
+// updates the toAddOrUpdateCache: after calling this function, the cache will only consist of sets to create
+// error handling principal:
+// - if contract with ipset save (or grep) is breaking, salvage what we can, take a snapshot without grep, and log the failure
+// - have a background process for sending/removing snapshots intermittently
+func (iMgr *IPSetManager) updateDirtyKernelSets(saveFile []byte, creator *ioutil.FileCreator) {
```
Comment (JungukCho) [inline npm/pkg/dataplane/ipsets/ipsetmanager_linux.go:205]: Does this function walk through all ipset information stored in kernel?
If so, is it necessary?
If there are large number of ipsets are programmed in kernel, this cost will be high?

Do we leverage `-exist` option for `add` and `del` to make it simple?
```shell
# from IPSet man page
add SETNAME ADD-ENTRY [ ADD-OPTIONS ]
Add a given entry to the set. If the -exist option is specified, ipset ignores if the entry already added to the set.
del SETNAME DEL-ENTRY [ DEL-OPTIONS ]
Delete an entry from a set. If the -exist option is specified and the entry is not in the set (maybe already expired), then the command is ignored.
```

File: npm/pkg/dataplane/ipsets/ipsetmanager_linux.go [hunk 2]
```diff
@@ -29,153 +67,409 @@ func (iMgr *IPSetManager) resetIPSets() error {
...
+		createMatches := nameForCreateRegex.FindSubmatch(createLine)
+		if len(createMatches) == lengthForSuccessfulMatch {
+			return
+		}
+	}
+	return
+}
+
+func haveTypeProblem(set *IPSet, restOfCreateLine []byte) bool {
+	// TODO check maxelem for hash sets? CIDR blocks have a different maxelem
```
Comment (JungukCho) [inline npm/pkg/dataplane/ipsets/ipsetmanager_linux.go:314]: I am not sure about `TODO`, but just context of maxelem for CIDR is the maximum number of ip address for the hashset -> 2^32 -1 -> 4,294,967,295 based on my understanding. This only works for `ipv4` though

File: npm/pkg/dataplane/policies/chain-management_linux.go [hunk 3]
```diff
@@ -90,9 +90,8 @@ func (pMgr *PolicyManager) initializeNPMChains() error {
 // removeNPMChains removes the jump rule from FORWARD chain to AZURE-NPM chain
 // and flushes and deletes all NPM Chains.
 func (pMgr *PolicyManager) removeNPMChains() error {
-	deleteErrCode, deleteErr := pMgr.runIPTablesCommand(util.IptablesDeletionFlag, jumpFromForwardToAzureChainArgs...)
-	hadDeleteError := deleteErr != nil && deleteErrCode != couldntLoadTargetErrorCode
-	if hadDeleteError {
+	deleteErrCode, deleteErr := pMgr.runIPTablesCommandAndIgnoreErrorCode(couldntLoadTargetErrorCode, util.IptablesDeletionFlag, jumpFromForwardToAzureChainArgs...)
```
Comment (JungukCho) [inline npm/pkg/dataplane/policies/chain-management_linux.go:93]: In deletion operation, is it ok not to get error code about `doesNotExistErrorCode` or `couldntLoadTargetErrorCode`, etc?
Anyway, the goal is to clean up the tables.

File: npm/pkg/dataplane/policies/chain-management_linux.go [hunk 4]
```diff
@@ -161,10 +164,11 @@ func (pMgr *PolicyManager) runIPTablesCommand(operationFlag string, args ...stri
 		errCode := exitError.ExitStatus()
 		allArgsString := strings.Join(allArgs, " ")
 		msgStr := strings.TrimSuffix(string(output), "\n")
-		if errCode > 0 && operationFlag != util.IptablesCheckFlag {
+		if errCode > 0 && errCode != errCodeToIgnore {
```
Comment (JungukCho) [inline npm/pkg/dataplane/policies/chain-management_linux.go:167]: Can you check `removeNPMChains` case with `errCodeToIgnore`?
In case the chain does not exists, it is fine to proceed.

File: npm/pkg/dataplane/policies/chain-management_linux.go [hunk 5]
```diff
@@ -146,6 +145,10 @@ func (pMgr *PolicyManager) reconcileChains(stopChannel <-chan struct{}) {
 
 // this function has a direct comparison in NPM v1 iptables manager (iptm.go)
 func (pMgr *PolicyManager) runIPTablesCommand(operationFlag string, args ...string) (int, error) {
+	return pMgr.runIPTablesCommandAndIgnoreErrorCode(-1, operationFlag, args...)
+}
+
+func (pMgr *PolicyManager) runIPTablesCommandAndIgnoreErrorCode(errCodeToIgnore int, operationFlag string, args ...string) (int, error) {
```
Comment (JungukCho) [inline npm/pkg/dataplane/policies/chain-management_linux.go:151]: It is a little hard to follow which operations are ok to ignore which errors.
Do we delegate error handling in a caller side?

We just have one `run` function and a caller will do right actions bases on return values.
It will be easier to follow up codes.

**Comment by JungukCho** [comment]
One comment: naming (file-creator.go and its contents) is misleading if I correctly understand the code. IPset and iptables use `stdout`  to program them. Initially, I am looking for where is file.

File: npm/pkg/dataplane/ipsets/ipsetmanager_linux.go (hunk 1 above)
Comment (huntergregory) [reply npm/pkg/dataplane/ipsets/ipsetmanager_linux.go:205]: we grep for all npm sets, so the work is linear in the number of sets in NPM. For non-dirty sets, we do skip reading all add lines in the save file, so this won't be too intensive. We wouldn't know what sets to delete without walking through the ipset save file, so `del --exist` couldn't work, and since we're already determining the members to delete for dirty sets, determining the exact members to add is equal work, and we write a smaller restore file compared to looping through all members and running `add --exist`

File: npm/pkg/dataplane/ipsets/ipsetmanager_linux.go [hunk 6]
```diff
@@ -29,153 +67,409 @@ func (iMgr *IPSetManager) resetIPSets() error {
...
+-A new-set-2 2.3.4.5
+-A new-set-1 2.3.4.5
+-A new-set-1 1.2.3.4
+-A new-set-3 1.2.3.4
+-A new-set-3 2.3.4.5
+
+save file for example above:
+	create set-in-kernel-1 net:hash ...
+	add set-in-kernel-1 1.2.3.4
+	add set-in-kernel-1 8.8.8.8
```
[the remark this answers is not in the document]
Comment (huntergregory) [reply npm/pkg/dataplane/ipsets/ipsetmanager_linux.go:125]: rearranging this for clarity

File: npm/pkg/dataplane/ipsets/ipsetmanager_linux.go (hunk 2 above)
Comment (huntergregory) [reply npm/pkg/dataplane/ipsets/ipsetmanager_linux.go:314]: clarified

File: npm/pkg/dataplane/ipsets/ipsetmanager_linux.go [hunk 7]
```diff
@@ -2,21 +2,59 @@ package ipsets
...
+	// constants for parsing ipset save
+	ipsetAddString           = "add"
+	ipsetCreateString        = "create"
+	ipsetSetListString       = "list:set"
+	ipsetNetHashString       = "hash:net"
+	ipsetIPPortHashString    = ipsetIPPortHashFlag
+	lengthForSuccessfulMatch = 2
+
+	// creator constants
+	maxTryCount                    = 3
```
[the remark this answers is not in the document]
Comment (huntergregory) [reply npm/pkg/dataplane/ipsets/ipsetmanager_linux.go:41]: to my knowledge, there are only line-based errors in ipset restore. all lines before an error do get applied so we handle the error at a given line by restoring from the line after the failure (and removing any subsequent, related lines). So if an IP was malformed and we missed it, then the add for it will fail, we'll continue on, and the everything else will be applied as it should be

File: npm/pkg/dataplane/ioutil/file-creator.go [hunk 8]
```diff
@@ -245,26 +261,24 @@ func (creator *FileCreator) handleLineError(lineNum int, commandString, stdErr s
...
-			if !exists {
-				log.Errorf("can't abort section because line references section %d which doesn't exist, so skipping the line instead", line.sectionID)
-				creator.lineNumbersToOmit[lineNumIndex] = struct{}{}
-			} else {
-				for _, lineNum := range section.lineNums {
-					creator.lineNumbersToOmit[lineNum] = struct{}{}
-				}
+		case Continue:
+			klog.Errorf("continuing after line %d for command [%s]", lineNum, commandString)
+			for k := 0; k <= lineNumIndex; k++ {
```
[the remark this answers is not in the document]
Comment (huntergregory) [reply npm/pkg/dataplane/ioutil/file-creator.go:266]: just habit haha I think I used to avoid i because it's the symbol for imaginary numbers. will change to i

[the remark this answers is not in the document]
Comment (huntergregory) [reply npm/pkg/dataplane/ioutil/file-creator.go:266]: just habit haha I think I used to avoid i because it's the symbol for imaginary numbers. will change to i

File: npm/pkg/dataplane/ioutil/grep.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+	commandError = command.Start()
+	if commandError != nil {
+		return
+	}
+
+	// Without this wait, defunct iptable child process are created
+	wait := func() { _ = command.Wait() } // appease go lint
+	defer wait()
+
+	output, err := grepCommand.CombinedOutput()
```
[the remark this answers is not in the document]
Comment (huntergregory) [reply npm/pkg/dataplane/ioutil/grep.go:26]: no actually since grep returns an error if we don't find anything. I'd like to return commandError nil for this case

File: npm/pkg/dataplane/policies/chain-management_linux.go (hunk 5 above)
Comment (huntergregory) [reply npm/pkg/dataplane/policies/chain-management_linux.go:151]: The main consideration for ignoring an error code in this function is that this function logs when there's an error, but I think it shouldn't log an error if we ignore the error later. We previously had some of that code baked into run() for check operations, but I think we should also do it for deleting a rule.

File: npm/pkg/dataplane/policies/chain-management_linux.go (hunk 4 above)
Comment (huntergregory) [reply npm/pkg/dataplane/policies/chain-management_linux.go:167]: sorry, could you clarify this?

File: npm/pkg/dataplane/policies/chain-management_linux.go (hunk 3 above)
Comment (huntergregory) [reply npm/pkg/dataplane/policies/chain-management_linux.go:93]: if the AZURE-NPM chain doesn't exist, we'll get couldntLoadTargetErrorCode and if the chain exists but the rule doesn't exist, we'll get doesNotExistErrorCode. So this should ignore both of these codes actually, since this is called before initializing DP.

**Comment by huntergregory** [comment]
> One comment: naming (file-creator.go and its contents) is misleading if I correctly understand the code. IPset and iptables use `stdout` to program them. Initially, I am looking for where is file.

Open to suggestions. How about restore-file-creator.go?

File: npm/pkg/dataplane/policies/chain-management_linux.go (hunk 3 above)
Comment (huntergregory) [reply npm/pkg/dataplane/policies/chain-management_linux.go:93]: if there's an error besides those two, something's wrong, but it could be several things, so ya maybe we just log and continue.

For instance, we get `couldntLoadTargetErrorCode` for "unknown option" and for a malformed IP address.

File: npm/pkg/dataplane/ioutil/file-creator.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (JungukCho) [reply npm/pkg/dataplane/ioutil/file-creator.go]: You are mathematician!

File: npm/pkg/dataplane/ipsets/ipsetmanager_linux.go (hunk 1 above)
Comment (JungukCho) [reply npm/pkg/dataplane/ipsets/ipsetmanager_linux.go:210]: I may misunderstand the code, but in the code, it also  has string match. So the work is not linear to the number of sets. It is linear to the number of sets and its string in each set.

My fundamental question in the code is why we need to walk through all ipsets (read from kernel and store in `saveFile` variable) in every `applyIPSets`  call and derive some actions (add or deletion) even though we maintain a lot of data structures in  `ipsetmanager_linux.go`. Can we leverage those information and apply what we need?

File: npm/pkg/dataplane/policies/chain-management_linux.go (hunk 4 above)
Comment (JungukCho) [reply npm/pkg/dataplane/policies/chain-management_linux.go]: Seems you resolved it.

**Comment by JungukCho** [comment]
> > One comment: naming (file-creator.go and its contents) is misleading if I correctly understand the code. IPset and iptables use `stdout` to program them. Initially, I am looking for where is file.
> 
> Open to suggestions. How about restore-file-creator.go?
May be restoreLinuxDP.go since both iptables and ipset uses this for mainly restore, but your suggestion is ok. 
Just having `file` makes me confused, but I may be so sensitive.

File: npm/pkg/dataplane/ipsets/ipsetmanager_linux.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (JungukCho) [reply npm/pkg/dataplane/ipsets/ipsetmanager_linux.go:41]: I am not sure, but failure will be cascading (For example, if `create` failed, adding elements will be failed). 

The other concern is that I am not sure it is easy to resolve failed one in a next reconcile.
If you simply handle only failed one in a next reconcile, I think it will be fine.

**Comment by huntergregory** [comment]
> > > One comment: naming (file-creator.go and its contents) is misleading if I correctly understand the code. IPset and iptables use `stdout` to program them. Initially, I am looking for where is file.
> > 
> > 
> > Open to suggestions. How about restore-file-creator.go?
> > May be restoreLinuxDP.go since both iptables and ipset uses this for mainly restore, but your suggestion is ok.
> > Just having `file` makes me confused, but I may be so sensitive.

ok I'll call it restore_linux.go. Are you ok with the name FileCreator for the struct?

File: npm/pkg/dataplane/ipsets/ipsetmanager_linux.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (huntergregory) [reply npm/pkg/dataplane/ipsets/ipsetmanager_linux.go:41]: the current logic will handle cascading errors (e.g. if a create fails, then abandon all adds/deletes for that set)

eventually we will save these failures in a reconcile cache and handle separately

File: npm/pkg/dataplane/ipsets/ipsetmanager_linux.go (hunk 1 above)
Comment (huntergregory) [reply npm/pkg/dataplane/ipsets/ipsetmanager_linux.go:210]: So I guess the decision is between these options:

1. read in save file for dirty sets
2. maintain copy of ipsets object from before they became dirty

We're doing the first here. The second would require extra memory (worst case, doubling the memory for the ipset cache before applying) and would require a redesign of the whole ipset manager. 

Thoughts @vakalapa ?

Comment (vakalapa) [reply npm/pkg/dataplane/ipsets/ipsetmanager_linux.go:210]: This operation is definitely very intensive wrt string matching, even if we skip over the section of some ipsets we will be doing first of line matching. 

We chose not to save deleted members of IPSets to save memory because we are already very constrained on memory in linux and data structures we are maintaining in Ipset manager is just source of truth. This is modelled against k8s philosophy, we have a source of truth (our internal ipset cache), we read the state on the machine and try to reconcile to that state.

Comment (vakalapa) [reply npm/pkg/dataplane/ipsets/ipsetmanager_linux.go:210]: Also, this approach is robust in the sense that kernel state will hardly deviate from expected state since NPM takes a snapshot of sets to be updated and then applies changes based on that, current IPSM's draw back is that as errors occur and as NPM runs for longer duration, kernel state heavily deviates from what is expected and this results in a cascading affect of issues, 

Hopefully in the future, when we figure out all the errors and have low-probability-for-error controller, then we can tone this approach down. But for initial purposes i feel, even with this compute intensive update, we are VERY fast compared to previous generation.

Wdygt?

Comment (JungukCho) [reply npm/pkg/dataplane/ipsets/ipsetmanager_linux.go:210]: If you guys think this is a way, I am ok with this approach.

I just wrote my two cents.
Based on my understanding, the goal of this enhancements in ipset in high-level is
1) Compared to v1 NPM, in v2 we do kind of batch process (one shot for one controller event) unlike in v1 NPM (sequentially called exec.run call for each ipset operation) to improve performance.
2) More fined-grained ipset management with reference counts to avoid known issues (e.g., `used in kernel`)
Do I miss something?

I may be wrong and miss some details,  but to me, it looks like that we took a kind of reverse engineering way, which is hard and complex way while we have enough information to derive all information since data plane is based on all information in control plane.
1) We know when and which operations we need to program (e.g., add, deletion, update) based on controller event) 
2) We know what information are programmed from ipset manger data-structures
3) All the references counts for ipsets
So, I think we can achieve the same functionality with those information in much simple way.

About performance, with a lot of ipsets, I am not sure how much we gains since we do string match O(# of ipsets * # of characters per ipset) in every event which needs applyIPset.

**Comment by JungukCho** [comment]
> > > > One comment: naming (file-creator.go and its contents) is misleading if I correctly understand the code. IPset and iptables use `stdout` to program them. Initially, I am looking for where is file.
> > > 
> > > 
> > > Open to suggestions. How about restore-file-creator.go?
> > > May be restoreLinuxDP.go since both iptables and ipset uses this for mainly restore, but your suggestion is ok.
> > > Just having `file` makes me confused, but I may be so sensitive.
> 
> ok I'll call it restore_linux.go. Are you ok with the name FileCreator for the struct?

No strong opinion.

File: npm/pkg/dataplane/ipsets/ipsetmanager_linux.go (hunk 1 above)
Comment (vakalapa) [reply npm/pkg/dataplane/ipsets/ipsetmanager_linux.go:210]: As we discussed offline, this current design is more secure than alternatives because even if some malicious actor changes the ipsets, NPM will try to reconcile to the desired state and overwrite manual entries. Granted NPM will only do this for "dirty" sets. A backlog item is added to have a background thread "at leisure" should try to reconcile all ipsets to the desired state, this design will make NPM more secure in the long run. 

Also, perf evaluation is planned on this current design to understand at scale what latency this compute intensive task is going to induce in rule programming.

File: npm/pkg/dataplane/policies/chain-management_linux.go (hunk 3 above)
Comment (huntergregory) [reply npm/pkg/dataplane/policies/chain-management_linux.go]: these changes will be moved to the next PR which updates policy/chain management

File: npm/pkg/dataplane/policies/chain-management_linux.go (hunk 4 above)
Comment (huntergregory) [reply npm/pkg/dataplane/policies/chain-management_linux.go]: these changes will be moved to the next PR which updates policy/chain management

File: npm/pkg/dataplane/policies/chain-management_linux.go (hunk 5 above)
Comment (huntergregory) [reply npm/pkg/dataplane/policies/chain-management_linux.go]: these changes will be moved to the next PR which updates policy/chain management

File: npm/config/config.go [hunk 10]
```diff
@@ -17,7 +17,7 @@ var DefaultConfig = Config{
 		EnablePrometheusMetrics: true,
 		EnablePprof:             true,
```
Comment (JungukCho) [inline npm/config/config.go]: Just check-in:
Based on our discussion, it would be better to set `EnablePprof` as `false`?

Comment (JungukCho) [reply npm/config/config.go]: Oops. Reverted! nevermind.

File: npm/pkg/dataplane/ipsets/ipsetmanager_linux.go [hunk 11]
```diff
@@ -29,153 +67,414 @@ func (iMgr *IPSetManager) resetIPSets() error {
...
-	creator := iMgr.getFileCreator(maxTryCount, toDeleteSetNames, toAddOrUpdateSetNames)
-	err := creator.RunCommandWithFile(util.Ipset, util.IpsetRestoreFlag)
-	if err != nil {
-		return fmt.Errorf("%w", err)
+	var saveFile []byte
+	var saveError error
+	if len(iMgr.toAddOrUpdateCache) > 0 {
+		saveFile, saveError = iMgr.ipsetSave()
+		if saveError != nil {
+			return fmt.Errorf("%w", saveError)
```
Comment (matmerr) [inline npm/pkg/dataplane/ipsets/ipsetmanager_linux.go:144]: can we wrap with context so we know the save error came from this stack

**Review by JungukCho — APPROVED** [review]
LGTM! Thank you for applying comments. 
We can walk about translation and linux dataplane before integration test.

**Review by vakalapa — APPROVED** [review]
